api/traces/v1/http.go

```diff
@@ -163,6 +165,46 @@ func NewV2Handler(read *url.URL, opts ...HandlerOption) http.Handler {
 	return r
 }
 
+// Parse a URL; if the URL includes `{tenant}` that portion will be replaced by the tenant
+func ExpandTemplatedUpstream(templateUpstream, tenant string) (*url.URL, error) {
+	rawTracesReadEndpoint := strings.Replace(templateUpstream, "{tenant}", tenant, 1)
+	return url.ParseRequestURI(rawTracesReadEndpoint)
+}
+
+// middlewareSetTemplatedUpstream is a variation of proxy.MiddlewareSetUpstream()
+// with additional processing if the upstream includes "{tenant}"
+func middlewareSetUpstream(logger log.Logger, read *url.URL, readTemplate string) proxy.Middleware {
+	if read != nil {
+		return proxy.MiddlewareSetUpstream(read)
+	}
+
+	// Cache upstream URLs to avoid re-parse on every read
+	templateToURL := map[string]*url.URL{}
+
+	return func(r *http.Request) {
+		tenant, ok := authentication.GetTenant(r.Context())
+		if !ok {
+			// At this point another middleware must have put the tenant into the context
+			level.Debug(logger).Log("msg", "Internal error; expected tenant in request context")
+		}
+
+		upstream, ok := templateToURL[tenant]
+		if !ok {
+			upstream, err := ExpandTemplatedUpstream(readTemplate, tenant)
```

Comment:
What is stopping this being called with tenant being an empty string and does it matter?

Reply:
- _main.go_ has `r.Use(authentication.WithTenantHeader(cfg.traces.tenantHeader, tenantIDs))` which prevents tenant from being empty
- It doesn't matter; suppose the template is `http://jaeger-{tenant}-query:16686`, it will fail with 502 BAD GATEWAY unless there is a host jaeger--query visible to DNS.